add microscheme

ardumont opened the pull request:
- Add microscheme derivation (depends on vim for build)
- Update avr-gcc url to the new one

(updated)

ardumont commented:
Hello,

As usual, if someone (with some free time on their hand :) could point me on how this can be improved, it would be great.

Cheers,

bjornfor commented:
Please squash the "Use git source code (tarball segfaults)" commit. PRs with fixes on its own code is just noise :-)

And in general, commits that touch specific packages should preferrably use a "pkgname: what" message format.

ardumont commented:
Thanks!

ardumont commented:
> Please squash the "Use git source code (tarball segfaults)" commit. PRs with fixes on its own code is just noise :-)

Yes, it's intended noise at first to explain to maintainers the choice made
(git over the `stable` tarball which did not work :D)

And now, it's squashed.

> And in general, commits that touch specific packages should preferrably use a "pkgname: what" message format.

Done!

Thanks for the feedback!

Cheers,

ardumont commented:
Rebased to latest upstream/master commit.

ardumont commented:
I am sorry but I am not able to determine if the CI broke because of this PR or not...

bjornfor commented:
> > Please squash the "Use git source code (tarball segfaults)" commit. PRs with fixes on its own
> > code is just noise :-)

> Yes, it's intended noise at first to explain to maintainers the choice made
> (git over the stable tarball which did not work :D)

Explaining choices is good idea.

But, IMHO, there is a difference between a (private) work-in-progress branch and a branch submitted for merging upstream. For the upstream project, it is more useful to see those explanations up front, instead of having them spread out in time and changing back and forth (possibly).

When you squash commits, you don't have to throw away good information such as "using git instead of tarball because tarball doesn't work". Just merge/edit the final commit message as you see fit.

ardumont commented:
> But, IMHO, there is a difference between a (private) work-in-progress branch and a branch submitted for merging upstream. For the upstream project, it is more useful to see those explanations up front, instead of having them spread out in time and changing back and forth (possibly).
> When you squash commits, you don't have to throw away good information such as "using git instead of tarball because tarball doesn't work". Just merge/edit the final commit message as you see fit.
 
Good points.
Thanks.

ardumont committed:
Fix avr-gcc url

New releases have been made and the current one is now an old one.

ardumont committed:
microscheme - Add microscheme derivation

Use of the actual git HEAD because the actual tarball compiles but
segfault at runtime.

Upload the BLINK.ms sample on arduino board (/dev/ttyACM0):

```sh
$ sudo ./result/bin/microscheme -m UNO -d /dev/ttyACM0 -auc ~/repo/perso/microscheme/examples/BLINK.ms
Microscheme 0.8, (C) Ryan Suchocki
>> Treeshaker: After 4 rounds: 84 globals purged! 22 bytes will be
reserved.
>> 18 lines compiled OK
>> Assembling...
>> Uploading...

avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100%
0.00s

avrdude: Device signature = 0x1e950f
avrdude: reading input file
"/home/tony/repo/perso/microscheme/examples/BLINK.hex"
avrdude: writing flash (2080 bytes):

Writing | ################################################## | 100%
0.35s

avrdude: 2080 bytes of flash written

avrdude: safemode: Fuses OK (E:00, H:00, L:00)

avrdude done.  Thank you.

>> Cleaning Up...
>> Finished.
```

ardumont commented:
Thanks.

> name should have a version number. If there is no version number corresponding to the git commit you're building from, use the date of the commit: "microscheme-YYYY-MM-DD".

Ok.

> It's considered better style to not pull in the whole pkgs set, but only the actually needed packages (i.e. 'vim').

Ok.

> There is no such thing as "fullDescription". It's called "longDescription".

Fixed.

> Please wrap overlong lines.

Done.

> Empty maintainers is of little use. Either add yourself or remove the line.

Right.
I added myself to the /lib/maintainers.nix.

jagajaga commented on pkgs/development/compilers/microscheme/default.nix:
```diff
@@ -0,0 +1,41 @@
+    inherit url;
+  };
+
+  buildInputs = [ pkgs.vim ];
```
Looks strange, it's ok? :)

ardumont replied:
> Looks strange, it's ok? :)

I initially created an `xxd` derivation to avoid this because it's really a deps on the `xxd` binary installed by vim.
The xxd derivation drew vim, went to the `src/xxd` folder, `make` the xxd...

@bjornfor suggested to avoid this and directly depend on vim because disk is cheap.

jagajaga commented:
IMO better squash all commits into one and add information in a commit message.

ardumont commented:
> IMO better squash all commits into one and add information in a commit message. 

I already did that.

@bjornfor then asked me to separate logically separated changes.
(quote: `This is a logically separate change, please put it in a separate commit.`)
Granted it initially referred to the update of the url of avr-gcc.

I'm a people of good will but If I keep doing what the last collaborator wants, I will keep doing and undoing what I have already done. 
Thus getting nowhere...

Cheers,

ardumont commented:
I squashed the last 2 commits.

bjornfor commented on pkgs/development/misc/avr-gcc-with-avr-libc/default.nix:
```diff
@@ -21,7 +21,7 @@ stdenv.mkDerivation {
 
   srcAVRLibc = fetchurl {
-    url = http://download.savannah.gnu.org/releases/avr-libc/avr-libc-1.7.1.tar.bz2;
+    url = http://ftp.igh.cnrs.fr/pub/nongnu/avr-libc/old-releases/avr-libc-1.7.1.tar.bz2;
```
Why change the mirror? All that's needed is adding ```old-releases/```.

ardumont replied:
I tested yesterday and somehow the result was not ok.
I may have been too prompt.
Good, I'll adapt.

ardumont replied:
Ok.
I confirm yesterday, my test was ko because I did not test the right url.
Here is the good one: http://download.savannah.gnu.org/releases/avr-libc/old-releases/avr-libc-1.7.1.tar.bz2

ardumont replied:
fix

ardumont commented:
To explicit one potential misreading.

> I'm a people of good will but If I keep doing `what the last collaborator wants`, I will keep doing and undoing what I have already done. 

When I say last collaborator, I did not mean you @jagajaga specifically, I meant generically the one after you then after her/him etc...

Cheers,

bjornfor commented:
Some more nitpicking:

"microscheme - Fix avr-gcc url which has changed." has little to do with microscheme. Please drop "microscheme" from that commit message.

Please add a comment why you're using git (HEAD?) instead of tarball. Tarballs (or *releases* really) are generally preferred (obviously).

Don't forget to update the microscheme expression to not pull in the whole nix package set (```pkgs```). Just add ```vim``` as input.

ardumont commented:
> Some more nitpicking:
> "microscheme - Fix avr-gcc url which has changed." has little to do with microscheme. Please drop "microscheme" from that commit message.

Ok.

> Please add a comment why you're using git (HEAD?) instead of tarball. Tarballs (or releases really) are generally preferred (obviously).

Ok.

> Don't forget to update the microscheme expression to not pull in the whole nix package set (pkgs). Just add vim as input.

Yes, I thought I did but realized I did not get it right the first time.
Now I do.

bjornfor commented:
Some more findings:

Homepage is http://microscheme.org/, not the gihub page.

How useable is microscheme without avr-gcc and avrdude? I guess at least avr-gcc is a needed at all times, right? If so, microscheme should be wrapped to include avr-gcc in PATH. (grep for wrapProgram in nixpkgs to see examples).

Bonus: send small patch to upstream microscheme developer(s) to improve the "make install" target:
 - don't use sudo (they even say in the README to install with "sudo make install", so there is definitely no need for sudo _inside_ the makefile)
 - don't hardcode install to /usr/local/bin; install to PREFIX instead (which can _default_ to /usr/local/bin)
 - install examples (right now one has to clone the repo anyway to fully test microscheme, so the nix package isn't a "complete" solution yet)

ardumont commented:
> Some more findings:
> 
> Homepage is http://microscheme.org/, not the gihub page.

I'm sorry, I preferred the direct access to github rather than a page referencing github as first page and the old one as second.
I'll fix.

> How useable is microscheme without avr-gcc and avrdude? I guess at least avr-gcc is a needed at all times, right? If so, microscheme should be wrapped to include avr-gcc in PATH. (grep for wrapProgram in nixpkgs to see examples).

It's not usable, they ask to install them along.

I believe in iterative steps.

First, to have something that we did not before.

```sh
nix-env -i microscheme avr-gcc avrdude
```

Then improve to have microscheme and the needed deps.

```sh
nix-env -i microscheme
```

Thanks for the pointer to `wrapProgram`.
I will need some time.

> Bonus: send small patch to upstream microscheme developer(s) to improve the "make install" target:
> - don't use sudo (they even say in the README to install with "sudo make install", so there is definitely no need for sudo inside the makefile)
> - don't hardcode install to /usr/local/bin; install to PREFIX instead (which can default to /usr/local/bin)
> - install examples (right now one has to clone the repo anyway to fully test microscheme, so the nix package isn't a "complete" solution yet)

Sure, I'll see what I can do.

Thanks.

Cheers,

bjornfor commented:
> > Homepage is http://microscheme.org/, not the gihub page.

> I'm sorry, I preferred the direct access to github rather than a page referencing
> github as first page and the old one as second. I'll fix.

The way I read it, the current homepage is only temporarily stripped down, because it will get rewritten. Or did I misunderstand? (microscheme.org isn't going away, is it?) The homepage points to the canonical github repo, and the github repo points to microscheme.org. This way I know that the nix package expression you're adding is the official repo, and not some arbitrary fork.

I think they document that you have to install avr-gcc and avrdude because they assume that the user is building from the git repo. This PR is about adding microscheme as a _package_, and as such, it should automatically pull in everything needed for it be useable out of the box. If not, humans have to do the work the package system.

ryansuchocki commented:
To clarify: avr-gcc and avrdude are not necessary to build microscheme, but they are necessary to use it. (Without them, microscheme will just compile programs to assembly.) They should definitely be prerequisites for any microscheme package.

Sorry about the unstable stable binaries!

ardumont commented:
Hello,

> The way I read it, the current homepage is only temporarily stripped down, because it will get rewritten. Or did I misunderstand?

Yes, I understood something along those lines too.

>  (microscheme.org isn't going away, is it?)

I do not think so.

> The homepage points to the canonical github repo, and the github repo points to microscheme.org. This way I know that the nix package expression you're adding is the official repo, and not some arbitrary fork.

Right. I already fixed it locally.
But I did not push because I have not yet wrapped (:D) my head around the `wrapProgram` yet.

> I think they document that you have to install avr-gcc and avrdude because they assume that the user is building from the git repo. This PR is about adding microscheme as a package, and as such, it should automatically pull in everything needed for it be useable out of the box. If not, humans have to do the work the package system.

Almost right.
It's not for compile but for runtime.
All the more reason to do as you propose!

> To clarify: avr-gcc and avrdude are not necessary to build microscheme, but they are necessary to use it. (Without them, microscheme will just compile programs to assembly.) They should definitely be prerequisites for any microscheme package.

This confirms @bjornfor's point. Thanks.

Cheers,

ardumont commented on pkgs/development/compilers/microscheme/default.nix:
```diff
@@ -0,0 +1,41 @@
+
+  buildPhase = ''
+    make hexify
+    make build
```
This can be simplify to `make` or `make all` too.
Fixed locally.

bjornfor replied:
If "make" is enough, you don't need custom buildPhase at all; the standard buildPhase does "make" for you :-)

ardumont replied:
Yes, indeed!

ardumont commented:
Yes, I have tried and produced an environment to show the shortcomings described here.

Arduino plugged and ready:
```sh
# tony at dagobah in ~/repo/perso/nixpkgs on git:add-microscheme x [18:47:26]
$ nix-shell -A microscheme
\[\][nix-shell:~/repo/perso/nixpkgs]$\[\] sudo ./result/bin/microscheme -m UNO -d /dev/ttyACM0 -auc ~/repo/perso/microscheme/examples/BLINK.ms
Microscheme 0.8, (C) Ryan Suchocki
>> Treeshaker: After 4 rounds: 84 globals purged! 22 bytes will be reserved.
>> 18 lines compiled OK
>> Assembling...
sh: avr-gcc: command not found
>> Warning: Command may have failed. (Exit code 32512)
sh: avr-objcopy: command not found
>> Warning: Command may have failed. (Exit code 32512)
>> Uploading...
sh: avrdude: command not found
>> Warning: Command may have failed. (Exit code 32512)
>> Cleaning Up...
>> Finished.
```

So now I have something to aim for!
(I did in thought but now it is in the realm of reality :)

*Note*
I had to clean up my env outside of nix-shell though... otherwise it worked...
```sh
nix-env -e avr-gcc-glibc avrdude
nix-store --delete result
```

Cheers,

ardumont commented:
So now, I have something that works inside the nix-shell using `wrapProgram` and `propagatedBuildInputs`:

```sh
# tony at dagobah in ~/repo/perso/nixpkgs on git:add-microscheme o [20:06:54]
$ nix-shell -A microscheme

\[\][nix-shell:~/repo/perso/nixpkgs]$\[\] sudo ./result/bin/microscheme -m UNO -d /dev/ttyACM0 -auc ~/repo/perso/microscheme/examples/BLINK.ms
[sudo] password for tony:
Microscheme 0.8, (C) Ryan Suchocki
>> Treeshaker: After 4 rounds: 84 globals purged! 22 bytes will be reserved.
>> 18 lines compiled OK
>> Assembling...
>> Uploading...

avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.01s

avrdude: Device signature = 0x1e950f
avrdude: reading input file "/home/tony/repo/perso/microscheme/examples/BLINK.hex"
avrdude: writing flash (2080 bytes):

Writing | ################################################## | 100% 0.35s

avrdude: 2080 bytes of flash written

avrdude: safemode: Fuses OK (E:00, H:00, L:00)

avrdude done.  Thank you.

>> Cleaning Up...
>> Finished.

```

But somehow if I install it:
```sh
nix-env -f . -i microscheme
```
It does not work.
(Reloading a new session terminal env and executing again also fails.)

The only thing I see about runtime dependencies on the manual is `propagatedBuildInputs`.

Quoting the [manual](http://nixos.org/nixpkgs/manual/):
```txt
propagatedBuildInputs

    Like buildInputs, but these dependencies are propagated:
    that is, the dependencies listed here are added to the buildInputs
    of any package that uses this package as a dependency. So if package
    Y has propagatedBuildInputs = [X], and package Z has buildInputs = [Y],
    then package X will appear in Z’s build environment automatically.
```

This does not help me regarded what I want. But, this helps:

```txt
3. In the fixup phase, it writes the propagated build inputs
(propagatedBuildInputs) to the file $out/nix-support/propagated-user-env-packages.
nix-env recursively installs all packages listed in this file when you install
a package that has it. This ensures that a Perl package can find its dependencies.
```

*Note* 
The file do exists locally after building the derivation (containing what I want avrdude and avr-gcc-libc). But somehow, this is not installed with nix-env (the way I asked anyway).

```sh
# tony at dagobah in ~/repo/perso/nixpkgs on git:add-microscheme o [19:55:40]
$ ll result
lrwxrwxrwx 1 tony users 66 Feb 15 19:26 result -> /nix/store/p7xk7n7cri1sj2wk40027a4gdzxc2w5j-microscheme-2015-02-04

# tony at dagobah in ~/repo/perso/nixpkgs on git:add-microscheme o [20:00:18]
$ ll result/
total 12K
dr-xr-xr-x 2 root nixbld 4.0K Jan  1  1970 bin
dr-xr-xr-x 2 root nixbld 4.0K Jan  1  1970 examples
dr-xr-xr-x 2 root nixbld 4.0K Jan  1  1970 nix-support

# tony at dagobah in ~/repo/perso/nixpkgs on git:add-microscheme o [20:00:20]
$ cat result/nix-support/propagated-native-build-inputs
/nix/store/khd56lwgqk5n1jm9y2gysabcp5n6qnb7-avrdude-6.1 /nix/store/xjxvfdx5awyjhv0zzah35h401f094fbv-avr-gcc-libc
```

So, do you see anything wrong in the [latest derivation](https://github.com/NixOS/nixpkgs/commit/6471617a96d8bec76e44ddb1fbe3d86d77937bb5)?

Thanks.

Cheers,

bjornfor commented:
About the propagatedBuildInputs issue, that may be due to incorrect/incomplete documentation. It works like documented for perl, python and other other interpreted languages, but not for compiled languages, AFAIK. So just change propagatedBuildInputs to buildInputs.

I guess it's wrapProgram that lacks some error checking and silently fails. It should be ```--prefix PATH : $prog/bin```. Mind the ':' with surrounding spaces.

ardumont commented:
> About the propagatedBuildInputs issue, that may be due to incorrect/incomplete documentation. It works like documented for perl, python and other other interpreted languages, but not for compiled languages, AFAIK. So just change propagatedBuildInputs to buildInputs.

Ok.

(As a side note, do you have some hints as to why differentiates behavior between compiled and interpreted languages?)

> I guess it's wrapProgram that lacks some error checking and silently fails. It should be --prefix PATH : $prog/bin. Mind the ':' with surrounding spaces.

Ok.
Fixed.

Thanks.

--

It still fails the same way (in all the possible combinations of the fixes you propose)

Still, `avr-gcc` and `avrdude` seems not to be live in the same profile generation as microscheme.
I guess it's possible to pollute one's environment with too many trials and error, don't you think?

--

I believe we need to keep the `propagatedBuildInputs` though:

Listing the closure of microscheme with installed microscheme with the `propagatedBuildInputs` on, we have the right dependencies:

```sh
# tony at dagobah in ~/repo/perso/nixpkgs on git:add-microscheme x [21:14:20]
$ nix-store -qR $(which microscheme)
/nix/store/khd56lwgqk5n1jm9y2gysabcp5n6qnb7-avrdude-6.1
/nix/store/xjxvfdx5awyjhv0zzah35h401f094fbv-avr-gcc-libc
/nix/store/x56lws7y1p511gmidm84vgyh59h8qnmc-microscheme-2015-02-04
```

Without, we do not have them.

But still somehow, does not work...
I will try to clean some more...

Thanks.

bjornfor commented:
> (As a side note, do you have some hints as to why differentiates behavior between
> compiled and interpreted languages?)

No, sorry. I don't know how it's implemented.

> I believe we need to keep the propagatedBuildInputs though:

No, it's not needed. Once the wrapProgram is correct, you will see avr-gcc and avrdude correctly listed as runtime dependencies with "nix-store -qR ./result" command.

ardumont commented:
> No, it's not needed. Once the wrapProgram is correct, you will see avr-gcc and avrdude correctly listed as runtime dependencies with "nix-store -qR ./result" command.

Ok.

--

At last, i have something.
I removed the `postInstall` phase and add the `wrapProgram` call inside the `installPhase`...

And indeed, I have exactly what you said.
Thanks.

Cheers,

ardumont commented:
I did look other derivations as suggested... At least the first [2 pages of results](https://github.com/NixOS/nixpkgs/search?p=2&q=wrapProgram&utf8=%E2%9C%93)... 

I do not say it's for all of them but what I saw is that the ones using `postInstall` do not define other phases. And other wrap their programs directly in the `installPhase`. 
So I give it a try and it worked.

ardumont commented:
Hello,

# Context

Everything is now ok.

The PR on microscheme has been merged so I updated this derivation with the latest code. 

# Steps

Installing the derivation in my user environment:

```sh
# tony at dagobah in ~/repo/perso/nixpkgs on git:add-microscheme x [19:21:38]
$ nix-env -f . -i microscheme
installing ‘microscheme-2015-02-04’
these derivations will be built:
/nix/store/y58vxwwjj9xg3a9s4g45igww00416h4i-microscheme-2015-02-04.drv
building path (s) ‘/nix/store/bj3n81zyq6v38d9ddhl6mhizw3licyi0-microscheme-2015-02-04’
unpacking sources
unpacking source archive /nix/store/kybbnwqzbbg4bc4h0hsz9ss99qkyy455-microscheme-2f14781
source root is microscheme-2f14781
patching sources
configuring
no configure script, doing nothing
building
make flags: SHELL=/nix/store/nynw3i86jq35hy8sqfxfnqrz2zdmz56s-bash-4.3-p33/bin/bash
building hexify
echo "// Hexified internal microscheme files." > src/assembly_hex.c
xxd -i src/preamble.s >> src/assembly_hex.c
echo "// Hexified internal microscheme files." > src/microscheme_hex.c
xxd -i src/primitives.ms >> src/microscheme_hex.c
xxd -i src/stdlib.ms >> src/microscheme_hex.c
building build
gcc -ggdb -std=gnu99 -Wall -Wextra -o microscheme src/*.c
installing
building install
install -m755 ./microscheme /nix/store/bj3n81zyq6v38d9ddhl6mhizw3licyi0-microscheme-2015-02-04/bin
post-installation fixup
patching ELF executables and libraries in /nix/store/bj3n81zyq6v38d9ddhl6mhizw3licyi0-microscheme-2015-02-04
/nix/store/bj3n81zyq6v38d9ddhl6mhizw3licyi0-microscheme-2015-02-04/bin/.microscheme-wrapped
/nix/store/bj3n81zyq6v38d9ddhl6mhizw3licyi0-microscheme-2015-02-04/bin/microscheme
not an ELF executable
gzipping man pages in /nix/store/bj3n81zyq6v38d9ddhl6mhizw3licyi0-microscheme-2015-02-04
stripping (with flags -S) in /nix/store/bj3n81zyq6v38d9ddhl6mhizw3licyi0-microscheme-2015-02-04/bin
strip:/nix/store/bj3n81zyq6v38d9ddhl6mhizw3licyi0-microscheme-2015-02-04/bin/microscheme: File format not recognized
patching script interpreter paths in /nix/store/bj3n81zyq6v38d9ddhl6mhizw3licyi0-microscheme-2015-02-04
building path (s) ‘/nix/store/bsas873iq41vrwg6dl1d0cjhhdwfxqpa-user-environment’
created 6348 symlinks in user environment
```

And looking at the runtime dependencies, we have what we want:
```sh
# tony at dagobah in ~/repo/perso/nixpkgs on git:add-microscheme o [19:23:59]
$ nix-store -qR $ (which microscheme)
...
/nix/store/khd56lwgqk5n1jm9y2gysabcp5n6qnb7-avrdude-6.1
/nix/store/jkfcv5fansqzixg27b5znr6qa5jf36qw-gnused-4.2.2
/nix/store/xjxvfdx5awyjhv0zzah35h401f094fbv-avr-gcc-libc
/nix/store/bj3n81zyq6v38d9ddhl6mhizw3licyi0-microscheme-2015-02-04
```

And indeed, uploading a sample blink.ms works:

```sh
# tony at dagobah in ~/repo/perso/nixpkgs on git:add-microscheme o [19:23:22]
$ microscheme -m UNO -d /dev/ttyACM0 -auc blink.ms
Microscheme 0.8, (C) Ryan Suchocki
>> Treeshaker: After 4 rounds: 84 globals purged! 22 bytes will be reserved.
>> 18 lines compiled OK
>> Assembling...
>> Uploading...

avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.01s

avrdude: Device signature = 0x1e950f
avrdude: reading input file "/home/tony/repo/perso/microscheme/examples/BLINK.hex"
avrdude: writing flash (2080 bytes):

Writing | ################################################## | 100% 0.35s

avrdude: 2080 bytes of flash written

avrdude: safemode: Fuses OK (E:00, H:00, L:00)

avrdude done.  Thank you.

>> Cleaning Up...
>> Finished.
```

Great news!

Cheers,

ardumont commented:
Do I need to do some other adaptation or is it ok?
Do I need to rebase?

bjornfor commented:
Great!

FYI, now that you are using nix variables with the wrapProgram call, you don't need to put the runtime deps into buildInputs. (The nix variables will be expanded into /nix/store/... paths before being passed to the shell.) It doesn't hurt anyone that they are there, but I just thought you should know.

We generally try to keep top-level $out pretty "minimal". By that I mean that the $out/examples/ directory better be moved to $out/share/microscheme/examples/ or something like that.

ardumont commented:
> FYI, now that you are using nix variables with the wrapProgram call, you don't need to put the runtime deps into buildInputs. (The nix variables will be expanded into /nix/store/... paths before being passed to the shell.) It doesn't hurt anyone that they are there, but I just thought you should know.

Ok.
I did not fully understood that and I'm glad you told me.
Thanks!

I will remove them if this is not needed, it's all the less to maintain.

>  We generally try to keep top-level $out pretty "minimal". By that I mean that the $out/examples/ directory better be moved to $out/share/microscheme/examples/ or something like that.

Ok.

I will also adapt accordingly.
Thanks.

ardumont commented:
> I will remove them if this is not needed, it's all the less to maintain.

Done.

> I will also adapt accordingly.

And done!

bjornfor commented:
Pushed to master!

(I couldn't build it at master branch due to libsigsegv failure (https://github.com/NixOS/nixpkgs/issues/6028). But ```nix-env -f . -qa '*' --meta --xml --drv-path --show-trace >/dev/null``` didn't show any errors.)

@ardumont: Thanks for taking all suggestions into account.

bjornfor commented:
Cherry-picked to release-14.12. Build tested before pushing.

ardumont commented:
Great news!

Thanks for the exchange @bjornfor, I learned a great deal with your suggestions.

Cheers,
